Refactoring/view and view model

Phangg opened the pull request:
## 개요 (이슈 번호 및 요약)
- #166 

## 변경 사항 (작업 내용)

### enum 일부 제외하고, Constants.swift 파일로 분리

### 구글 회원 탈퇴 로직 작성
- apple 과 비슷한 방식으로, google 유저인지 확인이 되면 google 재인증 / google 측에서 삭제 / 유저 탈퇴

### MyPageView 그룹 파일들 코드 수정
ChangeUserNameView
- 뷰 코드 내부에 있던 메서드 AuthViewModel 로 분리

NavigationLink 주석 코드 수정
- 주석 해제 후, 수정 및 연결

AlarmStoreView
- AlarmListContent 구조체 (뷰) private 접근 제어자 추가
- AlarmListContent 에서 NavigationLink 으로 PostDetailView 넘어갈 수 있도록 코드 수정

UserProfileView
- UserProfileKFImage 구조체 (뷰) private 접근 제어자 추가
- 뷰 내부에 있던 메서드 (포토픽커 관련) AuthViewModel 로 분리

AppServiceInfoView
- SafariView 구조체 (뷰) private 접근 제어자 추가

### MainViewModel 및 Liked 그룹 코드 수정
MainViewModel
- getHottestDrinks 와 getHottestPosts 메서드에서 query 를 collectionReference 로 타입 캐스팅하는 부분에서 crash 가 나는 문제 수정

FirestoreDrinkService / FirestorePostService
- MainViewModel 의 문제를 해결하기 위해 각각의 서비스에 있는 메서드를 일부 수정
- fetchPostCollection / fetchDrinkCollection 메서드에서 collectionReference 뿐 아니라,  query 도 파라미터로 받도록 함
( 옵셔널 & 기본값 nil )
- query 를 받아온 경우에는 query 에서 getDocuments 를 하고, 아닌 경우는 collection 에서 getDocuments 를 하도록 함.

DrinkViewModel
- isSearching 프로퍼티 기본값 오류 수정.
( true -> false )

Liked 관련 뷰 그룹
- LikedDrinkListContent / LikedPostGridContent 구조체 (하위 뷰) 에 private 접근제어자 설정

### Drink 모델에서 채택하는 프로토콜 수정
- Hashable 이 두번 채택 되어있었음 (struct 라인에서 1번, extension 라인에서 1번)
Equatable 을 채택하도록 수정

### AppCheck 오류 수정
AppCheck - firebase 에서 에러가 지속적으로 발생
- Project 의 Capability 에서 AppAttest 를 추가
- JUDA.entitlements 파일에서 AppCheck 부분 development 에서 production 으로 수정 ( firebase 공식문서 참고 )

### 알림 권한 허용 수정 
회원 가입 시, 알림 권한 허용을 MainView 에서 받고 있음.
- 기존에는 onAppear 로 받아오는 코드였으나, 화면이 disappear 된적이 없는 상황이라 코드가 실행되지 않았음.
- 따라서, onChange 에 로그인 상태 를 바라보도록 하여 알림 허용 받을 수 있도록 수정

### 회원가입 완료 버튼 로직 수정
- 회원 가입 시, 완료 버튼 누르면, signInDoneButtonTapped() 메서드 실행
- 메서드 내부에서 애플 / 구글 확인하여 각각에 맞는 로직 수행하도록 수정

### MyPageView 수정
- 내가 작성한 술상 없을 경우 텍스트 보여줄 수 있도록 수정

### 프로필 사진 변경 로직 수정
- 프로필 사진 수정 시, 기존에 프로필 이미지가 nil 일 수 있음.
- 그런 경우에는 파베 user 내부의 프로필 이미지 주소에도 데이터를 저장해 줄 수 있도록 함.

### 스플래쉬뷰에서 유저 데이터 받도록 수정
- 처음 앱 시작 시, 유저 데이터 받는 코드

### 유저 프로필 URL 업데이트 메서드
- 필드데이터 업데이트 시, URL 타입으로 보냈더니 오류가 났음 ( 파베 오류 )
- url.absoluteString 으로 String 으로 보내서 저장.

### AuthViewModel 수정
signInDoneButtonTapped 메서드
- 회원 가입 완료 시, 애플 or 구글 체크하여 그에 맞게 로직 수행하는 메서드

회원 탈퇴 시, isLoading 값 변경하는 부분 수정

updateUserProfileImageURL
- 유저 이미지가 없었다가, 생겼을 경우 url 업데이트 해주는 메서드

addUserDataToStore
- 유저 데이터 파베에 저장하는 메서드 수정

getCurrentUser 메서드 로직 수정
- currentUser 가 옵셔널이라, 병렬 처리 된 taskGroup 내부의 task 에서 currentUser 의 유저필드, 술상, 등등 에 값을 직접 넣어줄 수 없는 문제 발생
- 각 task 에서 반환 한다고 하면, 모든 값이 다 다른 타입을 갖고 있기에 enum 으로 통일
- task 의 클로저 값을 받아서 User 타입에 한번에 다 같이 넣어서 저장
```swift
// 데이터를 한번에 병렬로 받아오기 위해 사용하는 enum
private enum CurrentUserResult {
    case userField(UserField?)
    case posts([Post]?)
    case likedPosts([Post]?)
    case likedDrinks([Drink]?)
    case notifications([UserNotification]?)
}

// 현재 CurrentUser : User 가져오기
func getCurrentUser() async {
    var userFieldResult: UserField?
    var postsResult: [Post]?
    var likedPostsResult: [Post]?
    var likedDrinksResult: [Drink]?
    var notificationsResult: [UserNotification]?
    do {
        let uid = try checkCurrentUserID()
        await withTaskGroup(of: CurrentUserResult.self) { taskGroup in
            // 현재 유저 UserField 받아오기
            taskGroup.addTask { .userField(await self.getCurrentUserField(uid: uid)) }
            // 현재 유저 Posts 받아오기
            taskGroup.addTask { .posts(await self.getCurrentUserPosts(uid: uid)) }
            // 현재 유저 LikedPosts 받아오기
            taskGroup.addTask { .likedPosts(await self.getCurrentUserLikedPosts(uid: uid)) }
            // 현재 유저 LikedDrinks 받아오기
            taskGroup.addTask { .likedDrinks(await self.getCurrentUserLikedDrinks(uid: uid)) }
            // 현재 유저 Notifications 받아오기
            taskGroup.addTask { .notifications(await self.getCurrentUserNotifications(uid: uid)) }
            // taskGroup 종료 시, result 받아서 메서드 내부의 프로퍼티에 할당
            for await result in taskGroup {
                switch result {
                case .userField(let userField):
                    userFieldResult = userField
                case .posts(let posts):
                    postsResult = posts
                case .likedPosts(let likedPosts):
                    likedPostsResult = likedPosts
                case .likedDrinks(let likedDrinks):
                    likedDrinksResult = likedDrinks
                case .notifications(let notifications):
                    notificationsResult = notifications
                }
            }
            // 옵셔널 해제
            guard let userField = userFieldResult,
                  let posts = postsResult,
                  let likedPosts = likedPostsResult,
                  let likedDrinks = likedDrinksResult,
                  let notifications = notificationsResult else { return }
            // 유저에 데이터 값 할당
            currentUser = User(userField: userField,
                               posts: posts,
                               likedPosts: likedPosts,
                               likedDrinks: likedDrinks,
                               notifications: notifications)
        }
    } catch {
        showError = true
        errorMessage = error.localizedDescription
        print("error :: getCurrentUser", error.localizedDescription)
    }
}

// 현재 유저 UserField 받아오기
private func getCurrentUserField(uid: String) async -> UserField? {
    do {
        let userField = try await firebaseUserService.fetchUserFieldData(uid: uid)
        return userField
    } catch {
        errorMessage = error.localizedDescription
        print("error :: getCurrentUserField", error.localizedDescription)
        return nil
    }
}

// 현재 유저 Posts 받아오기
func getCurrentUserPosts(uid: String) async -> [Post]? {
    do {
        let posts = try await firebaseUserService.fetchUserWrittenPosts(uid: uid)
        return posts
    } catch {
        errorMessage = error.localizedDescription
        print("error :: getCurrentUserPosts", error.localizedDescription)
        return nil
    }
}

// 현재 유저 LikedPosts 받아오기
private func getCurrentUserLikedPosts(uid: String) async -> [Post]? {
    do {
        let likedPosts = try await firebaseUserService.fetchUserLikedPosts(uid: uid)
        return likedPosts
    } catch {
        errorMessage = error.localizedDescription
        print("error :: getCurrentUserLikedPosts", error.localizedDescription)
        return nil
    }
}

// 현재 유저 LikedDrinks 받아오기
private func getCurrentUserLikedDrinks(uid: String) async -> [Drink]? {
    do {
        let likedDrinks = try await firebaseUserService.fetchUserLikedDrink(uid: uid)
        return likedDrinks
    } catch {
        errorMessage = error.localizedDescription
        print("error :: getCurrentUserLikedDrinks", error.localizedDescription)
        return nil
    }
}

// 현재 유저 Notifications 받아오기
private func getCurrentUserNotifications(uid: String) async -> [UserNotification]? {
    do {
        let notifications = try await firebaseUserService.fetchUserNotifications(uid: uid)
        return notifications
    } catch {
        errorMessage = error.localizedDescription
        print("error :: getCurrentUserNotifications", error.localizedDescription)
        return nil
    }
}
```

Phangg committed:
[Edit] enum 코드 Constans.swift 파일로 분리

Phangg committed:
[Add] 구글 회원탈퇴 로직 작성

### 구글 탈퇴 로직 생성

### 기존
MyPageView - SettingView 에서 회원 탈퇴할 때,  AuthViewModel 의 deleteAppleAccount / deleteGoogleAccount 를 각각 호출하고 있었음.

### 변경 후
deleteAccount 룰 호출하면, provider 로 분기를 줘서 각각 애플과 구글 탈퇴 메서드를 진행.

Phangg committed:
[Edit] MyPagView 그룹 코드 수정 #152

ChangeUserNameView
- 뷰 코드 내부에 있던 메서드 AuthViewModel 로 분리

NavigationLink 주석 코드 수정
- 주석 해제 후, 수정 및 연결

AlarmStoreView
- AlarmListContent 구조체 (뷰) private 접근 제어자 추가
- AlarmListContent 에서 NavigationLink 으로 PostDetailView 넘어갈 수 있도록 코드 수정

UserProfileView
- UserProfileKFImage 구조체 (뷰) private 접근 제어자 추가
- 뷰 내부에 있던 메서드 (포토픽커 관련) AuthViewModel 로 분리

AppServiceInfoView
- SafariView 구조체 (뷰) private 접근 제어자 추가

Phangg committed:
[Edit] MainViewModel 및 Liked 그룹 코드 수정 #152

MainViewModel
- getHottestDrinks 와 getHottestPosts 메서드에서 query 를 collectionReference 로 타입 캐스팅하는 부분에서 crash 가 나는 문제 수정

FirestoreDrinkService / FirestorePostService
- MainViewModel 의 문제를 해결하기 위해 각각의 서비스에 있는 메서드를 일부 수정
- fetchPostCollection / fetchDrinkCollection 메서드에서 collectionReference 뿐 아니라,  query 도 파라미터로 받도록 함
( 옵셔널 & 기본값 nil )
- query 를 받아온 경우에는 query 에서 getDocuments 를 하고, 아닌 경우는 collection 에서 getDocuments 를 하도록 함.

DrinkViewModel
- isSearching 프로퍼티 기본값 오류 수정.
( true -> false )

Liked 관련 뷰 그룹
- LikedDrinkListContent / LikedPostGridContent 구조체 (하위 뷰) 에 private 접근제어자 설정

Phangg committed:
[Edit] Drink 모델 채택 프로토콜 수정 #166

Hashable 이 두번 채택 되어있었음 (struct 라인에서 1번, extension 라인에서 1번)
Equatable 을 채택하도록 수정

Phangg committed:
[Edit] AppCheck 오류 수정 #166

AppCheck - firebase 에서 에러가 지속적으로 발생
- Project 의 Capability 에서 AppAttest 를 추가
- JUDA.entitlements 파일에서 AppCheck 부분 development 에서 production 으로 수정 ( firebase 공식문서 참고 )

Phangg committed:
[Edit] 알림 권한 허용 수정 #166

회원 가입 시, 알림 권한 허용을 MainView 에서 받고 있음.
- 기존에는 onAppear 로 받아오는 코드였으나, 화면이 disappear 된적이 없는 상황이라 코드가 실행되지 않았음.
- 따라서, onChange 에 로그인 상태 를 바라보도록 하여 알림 허용 받을 수 있도록 수정

Phangg committed:
[Edit] 회원가입 완료 버튼 로직 수정 #166

회원 가입 시, 완료 버튼 누르면
signInDoneButtonTapped() 메서드 실행
- 메서드 내부에서 애플 / 구글 확인하여 각각에 맞는 로직 수행하도록 수정

Phangg committed:
[Edit] 마이페이지 뷰 수정 #166

내가 작성한 술상 없을 경우 텍스트 보여줄 수 있도록 수정

Phangg committed:
[Edit] 프로필 사진 변경 로직 수정 #166

프로필 사진 수정 시, 기존에 프로필 이미지가 nil 일 수 있음.
그런 경우에는 파베 user 내부의 프로필 이미지 주소에도 데이터를 저장해 줄 수 있도록 함.

Phangg committed:
[Edit] 스플래쉬뷰에서 유저 데이터 받도록 수정 #166

처음 앱 시작 시, 유저 데이터 받는 코드

Phangg committed:
[Edit] -

Phangg committed:
[Edit] 유저 프로필 URL 업데이트 메서드 #166

필드데이터 업데이트 시, URL 타입으로 보냈더니 오류가 났음 ( 파베 오류 )
- url.absoluteString 으로 String 으로 보내서 저장.

Phangg committed:
[Edit]  -

Phangg committed:
[Edit] AuthViewModel 수정 #166

signInDoneButtonTapped 메서드
- 회원 가입 완료 시, 애플 or 구글 체크하여 그에 맞게 로직 수행하는 메서드

회원 탈퇴 시, isLoading 값 변경하는 부분 수정

getCurrentUser 메서드 로직 수정
- currentUser 가 옵셔널이라, 병렬 처리 된 taskGroup 내부의 task 에서 currentUser 의 유저필드, 술상, 등등 에 값을 직접 넣어줄 수 없는 문제 발생
- 각 task 에서 반환 한다고 하면, 모든 값이 다 다른 타입을 갖고 있기에 enum 으로 통일
- task 의 클로저 값을 받아서 User 타입에 한번에 다 같이 넣어서 저장

updateUserProfileImageURL
- 유저 이미지가 없었다가, 생겼을 경우 url 업데이트 해주는 메서드

addUserDataToStore
- 유저 데이터 파베에 저장하는 메서드 수정

Phangg committed:
Merge branch 'develop' into refactoring/ViewAndViewModel

Phangg committed:
[Edit] pbx 오류 수정

withseon commented on JUDA_iOS/JUDA/View/Mypage/MypageDetail/AlarmStoreView.swift:
```diff
@@ -48,28 +48,24 @@ struct AlarmStoreView: View {
+                        post: alarm.likedPost,
+                        usedTo: .myPage)) {
+                            AlarmStoreListCell(alarm: alarm)
+                        }
```
유저 본인이 작성한 post에 대한 알림이니까 postUserType이 .reader가 아닌, .writer가 맞을 거 같습니다 ~!
이제 보니까 postUserType의 writter -> writer로 수정해야 할 거 같네요 .... 여태 알아차리지 못했어요 😂

Phangg replied:
맞네요..ㅜ reader 에서 writer 로 수정하면서 스펠링 잘못된것도 수정했어요!
꼼꼼하게 봐줘서 고마워요..!! 🙇‍♂️

withseon commented on JUDA_iOS/JUDA/ViewModel/Auth/AuthViewModel.swift:
```diff
@@ -131,81 +113,207 @@ final class AuthViewModel: ObservableObject {
+        return changeName.count >= 2 && changeName.count <= 10 && user.userField.name != changeName
+    }
+    
+    // 유저 프로핗 사진 변경 시, 사용되는 메서드
```
주석 오타났어요..!

withseon commented on JUDA_iOS/JUDA/Resource/Font +.swift:
```diff
@@ -56,7 +56,7 @@ extension Font {
 }
 
-enum FontType {
+private enum FontType {
```
이 파일은 생각치 못했는데... 꼼꼼함에 감탄합니다..!

withseon reviewed (commented):
역시 팡햄.. 꼼꼼하고 섬세해..... 고생하셨어요 -! 🥂

Phangg committed:
[Edit] AlarmStoreView 오류 수정 #166

본인 알림 리스트에 있는 술상에 접근 시, reader 가 아닌 writer 로 수정

추가로 PostUserType 의 writer case 오타 수정 (writter -> writer)

Phangg committed:
[Edit] 오타 수정

주석 오타 수정

bdrsky2010 commented on JUDA_iOS/JUDA/Resource/Components/CustomBottomSheet.swift:
```diff
@@ -7,12 +7,6 @@
-}
-
 // MARK: - .sheet 의 content 부분.
 struct CustomBottomSheetContent: View{
```
View{ <- 여기 한 칸 안띄어져있는거라면 띄어주십쇼..ㅎ

Phangg replied:
이런.. 말도안되는 일을 했네여....... 바로 수정 드갑니다..

bdrsky2010 commented on JUDA_iOS/JUDA/Resource/Constants.swift:
```diff
@@ -0,0 +1,196 @@
+    case liked
+    case myPage
+    case main
+}
```
여러 군데에서 사용하는 enum 따로 정리해두는 건 좋은 거 같습니다.

bdrsky2010 commented on JUDA_iOS/JUDA/View/Main/MainView.swift:
```diff
@@ -81,18 +79,19 @@ struct MainView: View {
-					}
-				}
+            }
+            .onChange(of: authViewModel.signInStatus) { newValue in
```
로그인했을 때 받아오는게 맞긴하네요. 수정 감사합니다.

bdrsky2010 reviewed (commented):
확인완료. 입니다

Phangg committed:
[Edit] whiteSpace 추가

Phangg committed:
Merge branch 'develop' into refactoring/ViewAndViewModel

Phangg committed:
[Edit] 병합 후 오류 수정
